{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yg/982c10113957_2gb06y92y35sx509h/T/ipykernel_11174/144752835.py:27: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  andme_data = pd.read_csv(andme_data_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def process_23andme_data(position_info_path, andme_data_path, output_path):\n",
    "\n",
    "    # Function to extract \"chrN:pos\"\n",
    "    def extract_chr_pos(value):\n",
    "        parts = value.split('_')\n",
    "        return f\"{parts[0]}:{parts[1]}\"\n",
    "        \n",
    "    def extract_alt_allele(value):\n",
    "        parts = value.split('_')\n",
    "        return f\"{parts[3]}\"\n",
    "    def extract_ref_allele(value):\n",
    "        parts = value.split('_')\n",
    "        return f\"{parts[2]}\"\n",
    "    \n",
    "        \n",
    "    def read_data(position_info_path, andme_data_path):\n",
    "        position_info = pd.read_csv(position_info_path)\n",
    "        # Create new column with \"chrN:pos\"\n",
    "        position_info['chr_pos'] = position_info['matching_columns'].apply(extract_chr_pos)\n",
    "        position_info['Alt'] = position_info['matching_columns'].apply(extract_alt_allele)\n",
    "        position_info['Ref'] = position_info['matching_columns'].apply(extract_ref_allele)\n",
    "\n",
    "\n",
    "        andme_data = pd.read_csv(andme_data_path, sep='\\t', comment='#', header=None, names=['rsid', 'chromosome', 'position', 'genotype'])\n",
    "        andme_data['chr_pos'] = andme_data.apply(lambda row: f\"chr{row['chromosome']}:{row['position']}\", axis=1)\n",
    "        return position_info, andme_data\n",
    "\n",
    "    def merge_data(position_info, andme_data):\n",
    "        return pd.merge(position_info, andme_data, left_on='chr_pos', right_on='chr_pos', how='left')\n",
    "\n",
    "    def create_allele_columns(data):\n",
    "        data['maternal_allele'] = data['genotype'].str[0]\n",
    "        data['paternal_allele'] = data['genotype'].str[1]\n",
    "        return data\n",
    "\n",
    "    def calc_dosage(allele, alt_allele):\n",
    "        return np.where(pd.isna(allele) | pd.isna(alt_allele), np.nan, np.where(allele == alt_allele, 1, 0))\n",
    "\n",
    "    def apply_dosage_calculation(data):\n",
    "        data['maternal_dosage'] = calc_dosage(data['maternal_allele'], data['Alt'])\n",
    "        data['paternal_dosage'] = calc_dosage(data['paternal_allele'], data['Alt'])\n",
    "        data['unphased_dosage'] = data['maternal_dosage'] + data['paternal_dosage']\n",
    "        return data\n",
    "\n",
    "    def create_phased_columns(data):\n",
    "        coord_prefix = data[\"chr_pos\"].str.split(\":\").str[0]\n",
    "        position = data[\"chr_pos\"].str.split(\":\").str[1]\n",
    "        ref = data['Ref'].astype(str)\n",
    "        alt = data['Alt'].astype(str)\n",
    "        data['phased_column_maternal'] = coord_prefix + '_' + position + '_' + ref + '_' + alt + '_maternal'\n",
    "        data['phased_column_paternal'] = coord_prefix + '_' + position + '_' + ref + '_' + alt + '_paternal'\n",
    "        data['unphased_column'] = coord_prefix + '_' + position + '_' + ref + '_' + alt + '_combined'\n",
    "\n",
    "        return data\n",
    "\n",
    "    def create_output_dataframe(data):\n",
    "        \n",
    "        new_data = {\n",
    "            **dict(zip(data[\"phased_column_maternal\"], data[\"maternal_dosage\"])),\n",
    "            **dict(zip(data[\"phased_column_paternal\"], data[\"paternal_dosage\"])),\n",
    "            **dict(zip(data[\"unphased_column\"], data[\"unphased_dosage\"]))\n",
    "        }\n",
    "        df = pd.DataFrame([new_data])\n",
    "        return df.sort_index(axis=1)\n",
    "\n",
    "    def save_to_csv(df, output_path):\n",
    "        df.to_csv(output_path, index=False)\n",
    "\n",
    "    # Pipeline execution\n",
    "    position_info, andme_data = read_data(position_info_path, andme_data_path)\n",
    "    merged_data = merge_data(position_info, andme_data)\n",
    "    merged_data = create_allele_columns(merged_data)\n",
    "    merged_data = apply_dosage_calculation(merged_data)\n",
    "    merged_data = create_phased_columns(merged_data)\n",
    "    output_df = create_output_dataframe(merged_data)\n",
    "    save_to_csv(output_df, output_path)\n",
    "    return output_df\n",
    "\n",
    "# Example usage:\n",
    "dat = process_23andme_data('../../Data/Filtered_raw_training_data_union/matching_columns_all.csv', \n",
    "                     '../../Data/23andMe_files/11703.23andme.9619.txt', \n",
    "                     'output.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.0\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "# Load and preprocess the data\n",
    "data_directory = '../../Data/Filtered_unphased_training_data_union_final/'\n",
    "\n",
    "\n",
    "\n",
    "for i in range(1, 23):\n",
    "    \n",
    "    file_name = data_directory + f\"23AndMe_PRS313_merged_chr{i}_matching_combined.parquet\"\n",
    "    data = pd.read_parquet(file_name)\n",
    "    \n",
    "    # Get all the chromosome 1 variants\n",
    "    chr = dat.filter(regex=f'^chr{i}_', axis=1)\n",
    "    print(len(data.columns) - chr.shape[1]/3)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RS_Number</th>\n",
       "      <th>Coord</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Distance</th>\n",
       "      <th>Dprime</th>\n",
       "      <th>R2</th>\n",
       "      <th>Correlated_Alleles</th>\n",
       "      <th>FORGEdb</th>\n",
       "      <th>RegulomeDB</th>\n",
       "      <th>Function</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15643</th>\n",
       "      <td>rs10493800</td>\n",
       "      <td>chr1:88177403</td>\n",
       "      <td>(G/A)</td>\n",
       "      <td>0.3295</td>\n",
       "      <td>20480.0</td>\n",
       "      <td>0.6157</td>\n",
       "      <td>0.0712</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15644</th>\n",
       "      <td>rs10922697</td>\n",
       "      <td>chr1:88127152</td>\n",
       "      <td>(T/C)</td>\n",
       "      <td>0.2250</td>\n",
       "      <td>-29771.0</td>\n",
       "      <td>0.8004</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15645</th>\n",
       "      <td>rs305445</td>\n",
       "      <td>chr1:88208135</td>\n",
       "      <td>(G/A)</td>\n",
       "      <td>0.2646</td>\n",
       "      <td>51212.0</td>\n",
       "      <td>0.8838</td>\n",
       "      <td>0.0259</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15646</th>\n",
       "      <td>rs12137116</td>\n",
       "      <td>chr1:88109828</td>\n",
       "      <td>(G/A)</td>\n",
       "      <td>0.3560</td>\n",
       "      <td>-47095.0</td>\n",
       "      <td>0.5007</td>\n",
       "      <td>0.0418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3a</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15647</th>\n",
       "      <td>rs4655854</td>\n",
       "      <td>chr1:88086894</td>\n",
       "      <td>(C/A)</td>\n",
       "      <td>0.4131</td>\n",
       "      <td>-70029.0</td>\n",
       "      <td>0.4678</td>\n",
       "      <td>0.0142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17677</th>\n",
       "      <td>rs115461551</td>\n",
       "      <td>chr1:120690218</td>\n",
       "      <td>(C/T)</td>\n",
       "      <td>0.1078</td>\n",
       "      <td>-597776.0</td>\n",
       "      <td>0.3968</td>\n",
       "      <td>0.1165</td>\n",
       "      <td>A=T,G=C</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17678</th>\n",
       "      <td>rs115706173</td>\n",
       "      <td>chr1:121280485</td>\n",
       "      <td>(A/G)</td>\n",
       "      <td>0.0048</td>\n",
       "      <td>-7509.0</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0295</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>&lt;CN2&gt;,&lt;CN3&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17679</th>\n",
       "      <td>rs79368528</td>\n",
       "      <td>chr1:120689085</td>\n",
       "      <td>(C/T)</td>\n",
       "      <td>0.0871</td>\n",
       "      <td>-598909.0</td>\n",
       "      <td>0.5304</td>\n",
       "      <td>0.1643</td>\n",
       "      <td>A=T,G=C</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17680</th>\n",
       "      <td>rs587774691</td>\n",
       "      <td>chr1:121137155</td>\n",
       "      <td>(A/G)</td>\n",
       "      <td>0.0028</td>\n",
       "      <td>-150839.0</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0172</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17681</th>\n",
       "      <td>rs79775675</td>\n",
       "      <td>chr1:120405606</td>\n",
       "      <td>(T/C)</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>-882388.0</td>\n",
       "      <td>0.4184</td>\n",
       "      <td>0.0104</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2039 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         RS_Number           Coord Alleles     MAF  Distance  Dprime      R2  \\\n",
       "15643   rs10493800   chr1:88177403   (G/A)  0.3295   20480.0  0.6157  0.0712   \n",
       "15644   rs10922697   chr1:88127152   (T/C)  0.2250  -29771.0  0.8004  0.0172   \n",
       "15645     rs305445   chr1:88208135   (G/A)  0.2646   51212.0  0.8838  0.0259   \n",
       "15646   rs12137116   chr1:88109828   (G/A)  0.3560  -47095.0  0.5007  0.0418   \n",
       "15647    rs4655854   chr1:88086894   (C/A)  0.4131  -70029.0  0.4678  0.0142   \n",
       "...            ...             ...     ...     ...       ...     ...     ...   \n",
       "17677  rs115461551  chr1:120690218   (C/T)  0.1078 -597776.0  0.3968  0.1165   \n",
       "17678  rs115706173  chr1:121280485   (A/G)  0.0048   -7509.0  1.0000  0.0295   \n",
       "17679   rs79368528  chr1:120689085   (C/T)  0.0871 -598909.0  0.5304  0.1643   \n",
       "17680  rs587774691  chr1:121137155   (A/G)  0.0028 -150839.0  1.0000  0.0172   \n",
       "17681   rs79775675  chr1:120405606   (T/C)  0.0096 -882388.0  0.4184  0.0104   \n",
       "\n",
       "      Correlated_Alleles  FORGEdb RegulomeDB Function Ref          Alt  \n",
       "15643                NaN      4.0          5      NaN   G            A  \n",
       "15644                NaN      3.0          7      NaN   T            C  \n",
       "15645                NaN      4.0          7      NaN   G            A  \n",
       "15646                NaN      6.0         3a      NaN   G            A  \n",
       "15647                NaN      2.0          5      NaN   C            A  \n",
       "...                  ...      ...        ...      ...  ..          ...  \n",
       "17677            A=T,G=C      6.0          7      NaN   C            T  \n",
       "17678                NaN      8.0          4      NaN   A  <CN2>,<CN3>  \n",
       "17679            A=T,G=C      4.0          5      NaN   C            T  \n",
       "17680                NaN      NaN          .      NaN   A            G  \n",
       "17681                NaN      4.0          7      NaN   T            C  \n",
       "\n",
       "[2039 rows x 13 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_variants = pd.read_csv('../../Data/Filtered_raw_training_data_union/23andMe_matching_variants_updated.csv')\n",
    "all_variants.loc[all_variants[\"Coord\"].str.contains(\"chr1:\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chr1_100443397_0        T\\n1        G\\n2        T\\n3        T\\n4        C\\n        ..\\n17677    C\\n17678    A\\n17679    C\\n17680    A\\n17681    T\\nName: Ref, Length: 17682, dtype: object_0                  C\\n1                  A\\n2                  G\\n3                  G\\n4                  A\\n            ...     \\n17677              T\\n17678    <CN2>,<CN3>\\n17679              T\\n17680              G\\n17681              C\\nName: Alt, Length: 17682, dtype: object_combined',\n",
       "       'chr1_100443397_0        T\\n1        G\\n2        T\\n3        T\\n4        C\\n        ..\\n17677    C\\n17678    A\\n17679    C\\n17680    A\\n17681    T\\nName: Ref, Length: 17682, dtype: object_0                  C\\n1                  A\\n2                  G\\n3                  G\\n4                  A\\n            ...     \\n17677              T\\n17678    <CN2>,<CN3>\\n17679              T\\n17680              G\\n17681              C\\nName: Alt, Length: 17682, dtype: object_maternal',\n",
       "       'chr1_100443397_0        T\\n1        G\\n2        T\\n3        T\\n4        C\\n        ..\\n17677    C\\n17678    A\\n17679    C\\n17680    A\\n17681    T\\nName: Ref, Length: 17682, dtype: object_0                  C\\n1                  A\\n2                  G\\n3                  G\\n4                  A\\n            ...     \\n17677              T\\n17678    <CN2>,<CN3>\\n17679              T\\n17680              G\\n17681              C\\nName: Alt, Length: 17682, dtype: object_paternal'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position_column = dat.columns[dat.columns.str.contains(\"10044339\")]\n",
    "position_column"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
